ipld/resolver/src/service.rs

```diff
@@ -1,17 +1,431 @@
+                    // All Client instances have been dropped.
+                    // We could keep the Swarm alive to serve content to others,
+                    // but we ourselves are unable to send requests. Let's treat
+                    // this as time to quit.
+                    None => { break; }
+                }
+            };
+        }
+        Ok(())
     }
+
+    /// Handle events that the [`NetworkBehaviour`] for our [`Behaviour`] macro generated, one for each field.
+    fn handle_behaviour_event(&mut self, event: BehaviourEvent<P>) {
+        match event {
+            BehaviourEvent::Ping(e) => self.handle_ping_event(e),
+            BehaviourEvent::Identify(e) => self.handle_identify_event(e),
+            BehaviourEvent::Discovery(e) => self.handle_discovery_event(e),
+            BehaviourEvent::Membership(e) => self.handle_membership_event(e),
+            BehaviourEvent::Content(e) => self.handle_content_event(e),
+        }
+    }
+
+    // Copied from Forest.
+    fn handle_ping_event(&mut self, event: ping::Event) {
+        let peer_id = event.peer.to_base58();
+        match event.result {
+            Ok(ping::Success::Ping { rtt }) => {
+                trace!(
+                    "PingSuccess::Ping rtt to {} is {} ms",
+                    peer_id,
+                    rtt.as_millis()
+                );
+            }
+            Ok(ping::Success::Pong) => {
+                trace!("PingSuccess::Pong from {peer_id}");
+            }
+            Err(ping::Failure::Timeout) => {
+                debug!("PingFailure::Timeout from {peer_id}");
+            }
+            Err(ping::Failure::Other { error }) => {
+                warn!("PingFailure::Other from {peer_id}: {error}");
+            }
+            Err(ping::Failure::Unsupported) => {
+                warn!("Banning peer {peer_id} due to protocol error");
+                self.swarm.ban_peer_id(event.peer);
+            }
+        }
+    }
+
+    fn handle_identify_event(&mut self, event: identify::Event) {
+        if let identify::Event::Error { peer_id, error } = event {
+            warn!("Error identifying {peer_id}: {error}")
+        }
+    }
+
+    fn handle_discovery_event(&mut self, event: discovery::Event) {
+        match event {
+            discovery::Event::Added(peer_id, _) => self.membership_mut().set_routable(peer_id),
+            discovery::Event::Removed(peer_id) => self.membership_mut().set_unroutable(peer_id),
+            discovery::Event::Connected(_, _) => {}
+            discovery::Event::Disconnected(_, _) => {}
+        }
+    }
+
+    fn handle_membership_event(&mut self, event: membership::Event) {
+        match event {
+            membership::Event::Skipped(peer_id) => {
+                // Don't repeatedly look up peers we can't add to the routing table.
+                if self.background_lookup_filter.insert(&peer_id) {
+                    self.discovery_mut().background_lookup(peer_id)
```

Comment:
Are we refreshing the bloom filter after some time? I don´t know if having a lot of peers after some point could lead to skipping legit peers.

Reply:
I am also thinking about this, but I haven't come up with a nice way to do it yet. A fresh start every now and then seems like a blunt and easy way, but leads to a burst of lookups after. Maybe a better protection would be to use a cache like the `guava` library supported in the JVM: it has limits on memory and a TTL as well.